tachyon/math/elliptic_curves/bn/bn_curve.h

```diff
@@ -22,6 +44,80 @@ class BNCurve {
     return f;
   }
 
+  // TODO(chokobole): Leave a comment to help understand readers.
+  template <typename G1AffinePointContainer, typename G2PreparedContainer>
+  static Fp12Ty MultiMillerLoop(const G1AffinePointContainer& a,
+                                const G2PreparedContainer& b) {
+    using EllCoeffs = typename bn::G2Prepared<Config>::EllCoeffs;
+
+    class Pair {
+     public:
+      Pair() = default;
+      Pair(const G1AffinePointTy* g1, const EllCoeffs* ell_coeffs)
+          : g1_(g1), ell_coeffs_(ell_coeffs) {}
+
+      const G1AffinePointTy& g1() const { return *g1_; }
+      const EllCoeff<Fp2Ty>& NextEllCoeff() const {
+        return (*ell_coeffs_)[idx_++];
+      }
+
+     private:
+      const G1AffinePointTy* g1_ = nullptr;
+      const EllCoeffs* ell_coeffs_ = nullptr;
+      mutable size_t idx_ = 0;
+    };
+
+    size_t size = std::size(a);
+    CHECK_EQ(size, std::size(b));
+    std::vector<Pair> pairs;
+    pairs.reserve(size);
+    for (size_t i = 0; i < size; ++i) {
+      if (!a[i].infinity() && !b[i].infinity()) {
+        pairs.emplace_back(&a[i], &b[i].ell_coeffs());
+      }
+    }
+
+    auto callback = [](absl::Span<const Pair> pairs) {
+      Fp12Ty f = Fp12Ty::One();
+      for (size_t i = std::size(Config::kAteLoopCount) - 1; i >= 1; --i) {
+        if (i != std::size(Config::kAteLoopCount) - 1) {
+          f.SquareInPlace();
+        }
+
+        for (const Pair& pair : pairs) {
+          Ell(f, pair.NextEllCoeff(), pair.g1());
+        }
+
+        int8_t bit = Config::kAteLoopCount[i - 1];
+        if (bit == 1 || bit == -1) {
+          for (const Pair& pair : pairs) {
+            Ell(f, pair.NextEllCoeff(), pair.g1());
+          }
+        }
+      }
+      return f;
+    };
+
+    std::vector<Fp12Ty> results =
+        base::ParallelizeMapByChunkSize(pairs, 4, callback);
```

Comment:
Is there any reason to fix the chunk size to 4?

Reply:
It's a constant! https://github.com/arkworks-rs/algebra/blob/ee0a4078b67037b4c4d22d330113a5d7c28fd6ce/ec/src/models/bn/mod.rs#L66